opendevin/runtime/plugins/jupyter/__init__.py

```diff
@@ -12,3 +19,57 @@ class JupyterRequirement(PluginRequirement):
     )  # The directory of this file (opendevin/runtime/plugins/jupyter)
     sandbox_dest: str = '/opendevin/plugins/jupyter'
     bash_script_path: str = 'setup.sh'
+
+    # ================================================================
+    # Plugin methods, which will ONLY be used in the runtime client
+    # running inside docker
+    # ================================================================
+
+
+class JupyterPlugin(Plugin):
+    name: str = 'jupyter'
+
+    def initialize(self, kernel_id: str = 'opendevin-default'):
+        self.kernel_gateway_port = find_available_tcp_port()
+        self.kernel_id = kernel_id
+        self.gateway_process = subprocess.Popen(
+            [
+                '/opendevin/miniforge3/bin/mamba',
+                'run',
+                '-n',
+                'base',
+                'poetry',
+                'run',
+                'jupyter',
+                'kernelgateway',
+                '--KernelGatewayApp.ip=0.0.0.0',
+                f'--KernelGatewayApp.port={self.kernel_gateway_port}',
+            ],
+            stderr=subprocess.STDOUT,
+        )
+        # read stdout until the kernel gateway is ready
+        while True and self.gateway_process.stdout is not None:
+            line = self.gateway_process.stdout.readline().decode('utf-8')
+            if 'at' in line:
```

Comment:
Little confuse about `at` 🤔

Reply:
I remember there was a sentence that is produced whenever the gateway server successfully run that include certain phrases. Before we pin down the exact version of the gateway using `.lock`, there was some variation in that sentence - and the only common word left is "at" :(

Will try to see if i can make it more informative.

Reply:
Haha, get it.